keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,559 @@
+a few drawbacks in following use-cases:
+
+1. An `initContainer` creates some configuration that needs to be consumed as an
+   environment variable by the main container. Existing mechanisms necessitate
+   additional ConfigMap or Secret API calls.
+
+2. when each k8s Node is initialized with an environment file that is used by
+   Pods running on that node, k8s can provide a mechanism to instantiate env
+   vars from this file.
+
+3. The user is using a container offered by a vendor that requires configurating
+   environment variables (for eg, license key, one-time secret tokens). They can
+   use an `initContainer` to fetch this info into a file and the main container
+   can consume.
+
+The syntactic enhancement proposed in this KEP offers a simplified method for
+populating environment variables directly from files. This change would
+streamline the process for users.
+
+### Goals
+
+1. Support instantiating a container's environment variables from a file. This
+file can be in any volume (hostPath, emptyDir or PVC) as long as the kubelet
+can access the file and parse it. kubelet will instantiate the env vars in the
+container from the specified file but it will not mount the file.
+
+2. All containers (container, initContainer, sidecar and ephemeral container)
+will be able to load env vars from a file.
+
+### Non-Goals
+
+1. We do not plan to support expansion of environment variables, that is, when an
+environment variable refers to another.
+
+2. We do not intend to update env vars if the file content has changed after the
+container consuming it has started.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+The user is using a container offered by a vendors that requires configuring
+env vars. [Some
+vendors](https://developer.hashicorp.com/vault/docs/platform/k8s/injector/examples#environment-variable-example)
+have to couple their app with Kubernetes to allow setting env vars in their
+container.
+
+Instead, k8s can provide a new mechanism that will be used to instantiate
+env vars from the specified file. The user can generate an env file using an
+`initContainer`. The container can then use the `fileRef` field to reference
+the file from which it should initialize the env vars. Behind the scene, kubelet will
+parse the env file and instantiate these env vars during container creation. It
+does not mount this file onto the main container. In the example below,
+the Pod's output includes `CONFIG_VAR=HELLO`.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test-pod
+spec:
+  initContainers:
+  - name: setup-envfile
+    image: registry.k8s.io/busybox
+    command: ['sh', '-c', 'echo "CONFIG_VAR: HELLO" > /etc/config/config.yaml']
+    volumeMounts:
+    - name: data
+      mountPath: /etc/config
+  containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    envFrom:
+    - fileRef:
+        path: config.yaml
```

Comment:
I maybe mistakem, but does this need the full path `/etc/config/config.yaml` ?

Reply:
`/etc/config` is the mountpoint of volume `data` in initContainer `setup-envfile`.  `/etc/config/config.yaml` is the absolute filepath for initContainer. 

When we are referencing this file in the main container `use-envfile`, we want to tell kubelet the path inside the filesystem of volume `data`, which is just `config.yaml`.